dnssec.go

```diff
@@ -301,16 +308,26 @@ func (rr *RRSIG) Sign(k crypto.Signer, rrset []RR) error {
 		return ErrAlg
 	}
 
-	h := hash.New()
-	h.Write(signdata)
-	h.Write(wire)
+	switch rr.Algorithm {
+	case ED25519:
+		signature, err := sign(k, append(signdata, wire...), crypto.Hash(0), rr.Algorithm)
```

Comment:
I haven't read the rfc/draft. Why does this need to be done like this? (some comments maybe?).
What's crypto.Hash(0)? And why can't we do the double `h.Write`?

Reply:
@miekg That's not directly related to the draft but to how ed25519 operates. Most signature schemes operate on hashed data while ed25519 performs the hashing internally and thus requires the original message. While there is a variant of ed25519 that uses pre-hashing, RFC 8080 does not use it – very few things do. Basically ed25519 needs to be passed the raw message without any hashing so it's special cased here.

The crypto.Hash(0) is basically used as a placeholder to mean 'not hashed', it's checked in [`(ed25519.PrivateKey).Sign`](https://github.com/golang/crypto/blob/bd6f299fb381e4c3393d1c4b1f0b94f5e77650c8/ed25519/ed25519.go#L55-L57).

It might be clearer if I switch this to calling [ed25519.Sign](https://github.com/golang/crypto/blob/bd6f299fb381e4c3393d1c4b1f0b94f5e77650c8/ed25519/ed25519.go#L96) directly.

Reply:
@miekg Scratch calling `ed25519.Sign` directly, that would prevent the use of opaque `crypto.Signers`. I'll improve the documentation here instead.